### PR TITLE
feat(gateway): enhance kube-auth-proxy TLS configuration based on APIServer settings

### PR DESCRIPTION
**Pull request by sukumars321**
## Description
- Updated the gateway controller to reconcile when the APIServer's TLS security profile changes, ensuring kube-auth-proxy deployment reflects the latest TLS settings.
- Added functionality to retrieve TLS minimum version and cipher suites from the APIServer configuration.
- Introduced tests to validate that kube-auth-proxy deployment arguments match the APIServer's TLS profile.
- Updated related documentation to include new requirements for APIServer configurations.

This change improves security by ensuring that the kube-auth-proxy adheres to the specified TLS settings of the cluster's APIServer.

https://redhat.atlassian.net/browse/RHOAIENG-61063

## How Has This Been Tested?

Tested the changes using `quay.io/rh-ee-suksubra/opendatahub-operator:v3.3.0-rhoaieng-61063` 
## Screenshot or short clip
https://github.com/user-attachments/assets/0d20d22b-0b76-4a72-a47e-e064dd1b07fb

## Merge criteria

- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has run the integration test pipeline and verified that it passed successfully
- [ ] New RELATED_IMAGE mappings are listed in ODH-Build-Config and RHOAI-Build-Config (CI validates names against build-config repos). Include links to build-config PRs in the description.

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [ ] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

## Summary by CodeRabbit

* **New Features**
  * Authentication proxy now respects cluster TLS security profile (min TLS version and cipher suites) and updates automatically when the profile changes.

* **Tests**
  * Added unit, integration, and e2e tests validating TLS profile translation and dynamic updates of the auth proxy.

* **Documentation**
  * Expanded E2E prerequisite guidance to include additional APIServer/TLS config lookups (minimal stubs are sufficient).

* **Chores**
  * Bumped several module dependencies; added read access to cluster APIServer for TLS lookups.

### CODE REVIEW

File: internal/controller/services/gateway/gateway_integration_helpers_test.go [hunk 1]
```diff
@@ -1190,6 +1201,71 @@ func RunSpecMutationCookieConfigTest(t *testing.T, setup TestSetup, cookieUpdate
...
 
+// RunAPIServerTLSProfileMutationTest patches the cluster APIServer TLS profile and validates kube-auth-proxy Deployment TLS args update.
+func RunAPIServerTLSProfileMutationTest(t *testing.T, setup TestSetup) {
+	g := NewWithT(t)
+	defer setup.Setup(t)()
+
+	intermediateSpec := configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
+	intermediateMinArg := fmt.Sprintf("--tls-min-version=%s", gateway.TLSMinVersionToName(intermediateSpec.MinTLSVersion))
+	intermediateCiphersArg := fmt.Sprintf("--tls-cipher-suite=%s",
+		strings.Join(ocpcrypto.OpenSSLToIANACipherSuites(intermediateSpec.Ciphers), ","))
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_integration_helpers_test.go]:
```suggestion
		gateway.TLSCipherSuitesFromProfileSpec(intermediateSpec))
```

thoughts?

File: internal/controller/services/gateway/gateway_tls.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+func TLSMinVersionToName(minVersion configv1.TLSProtocolVersion) string {
+	switch minVersion {
+	case configv1.VersionTLS12:
+		return "TLS1.2"
+	case configv1.VersionTLS13:
+		return "TLS1.3"
+	default:
+		// TLS1.2 is the default minimum version supported by the kube-auth-proxy
+		return "TLS1.2"
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls.go]:
```suggestion
	default:
		// TLS1.2 is the default minimum version supported by the kube-auth-proxy
		log.Info("MinTLSVersion not supported by kube-auth-proxy, using TLS1.2 as floor",
    "configuredVersion", string(minVersion))
		return "TLS1.2"
```

File: internal/controller/services/gateway/gateway_integration_helpers_test.go [hunk 3]
```diff
@@ -1190,6 +1201,71 @@ func RunSpecMutationCookieConfigTest(t *testing.T, setup TestSetup, cookieUpdate
...
+	expectedCiphersArg := fmt.Sprintf("--tls-cipher-suites=%s",
+		strings.Join(ocpcrypto.OpenSSLToIANACipherSuites(oldSpec.Ciphers), ","))
+
+	apiServer := &configv1.APIServer{}
+	g.Expect(setup.TC.K8sClient.Get(setup.TC.Ctx, types.NamespacedName{Name: cluster.ClusterAPIServerObj}, apiServer)).To(Succeed())
+	apiServer.Spec.TLSSecurityProfile = &configv1.TLSSecurityProfile{
+		Type: configv1.TLSProfileOldType,
+	}
+	g.Expect(setup.TC.K8sClient.Update(setup.TC.Ctx, apiServer)).To(Succeed())
+
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply internal/controller/services/gateway/gateway_integration_helpers_test.go]: @sukumars321 Please check this comment

File: internal/controller/services/gateway/gateway_integration_helpers_test.go [hunk 4]
```diff
@@ -1190,6 +1201,71 @@ func RunSpecMutationCookieConfigTest(t *testing.T, setup TestSetup, cookieUpdate
...
+		for _, arg := range args {
+			if arg == intermediateMinArg {
+				hasMin = true
+			}
+			if arg == intermediateCiphersArg {
+				hasCiphers = true
+			}
+		}
+		return hasMin && hasCiphers
+	}, TestTimeout, TestInterval).Should(BeTrue(), "Deployment should have Intermediate TLS args before APIServer mutation")
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply internal/controller/services/gateway/gateway_integration_helpers_test.go]: @sukumars321 Please check this

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+		})
+	}
+}
+
+func TestKubeAuthProxyTLSFromProfile(t *testing.T) {
+	t.Parallel()
+
+	intermediateSpec := configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
+	expectedMin := TLSMinVersionToName(intermediateSpec.MinTLSVersion)
+	expectedCiphers := strings.Join(ocpcrypto.OpenSSLToIANACipherSuites(intermediateSpec.Ciphers), ",")
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls_test.go]:
```suggestion
	expectedCiphers := TLSCipherSuitesFromProfileSpec(intermediateSpec)
```

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+	minVersion, cipherSuites := KubeAuthProxyTLSFromProfile(nil)
+	assert.Equal(t, expectedMin, minVersion)
+	assert.Equal(t, expectedCiphers, cipherSuites)
+	assert.Equal(t, "TLS1.2", minVersion)
+	assert.Contains(t, cipherSuites, "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256")
+
+	oldSpec := configv1.TLSProfiles[configv1.TLSProfileOldType]
+	minVersion, cipherSuites = KubeAuthProxyTLSFromProfile(&configv1.TLSSecurityProfile{Type: configv1.TLSProfileOldType})
+	assert.Equal(t, TLSMinVersionToName(oldSpec.MinTLSVersion), minVersion)
+	assert.Equal(t, strings.Join(ocpcrypto.OpenSSLToIANACipherSuites(oldSpec.Ciphers), ","), cipherSuites)
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls_test.go]:
```suggestion
	assert.Equal(t, TLSCipherSuitesFromProfileSpec(oldSpec), cipherSuites)
```

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+func TestGetKubeAuthProxyTLSFromAPIServer(t *testing.T) {
+	t.Parallel()
+
+	scheme := runtime.NewScheme()
+	require.NoError(t, configv1.Install(scheme))
+
+	intermediateSpec := configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
+	expectedMin := TLSMinVersionToName(intermediateSpec.MinTLSVersion)
+	expectedCiphers := strings.Join(ocpcrypto.OpenSSLToIANACipherSuites(intermediateSpec.Ciphers), ",")
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls_test.go]:
```suggestion
	expectedCiphers := TLSCipherSuitesFromProfileSpec(intermediateSpec)
```

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+					Type: configv1.TLSProfileOldType,
+				},
+			},
+		}
+		cli := fake.NewClientBuilder().WithScheme(scheme).WithObjects(apiServer).Build()
+
+		minVersion, cipherSuites, err := getKubeAuthProxyTLSFromAPIServer(context.Background(), cli)
+		require.NoError(t, err)
+		assert.Equal(t, TLSMinVersionToName(oldSpec.MinTLSVersion), minVersion)
+		assert.Equal(t, strings.Join(ocpcrypto.OpenSSLToIANACipherSuites(oldSpec.Ciphers), ","), cipherSuites)
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls_test.go]:
```suggestion
		assert.Equal(t, TLSCipherSuitesFromProfileSpec(oldSpec), cipherSuites)
```

File: internal/controller/components/modelsasservice/modelsasservice_support_test.go [hunk 9]
```diff
@@ -12,7 +12,7 @@ import (
 
 var rf = provider.NewDefaultDepProvider().GetResourceFactory()
 
-func buildTestResMap(t *testing.T, yamls ...string) resmap.ResMap { //nolint:ireturn
+func buildTestResMap(t *testing.T, yamls ...string) resmap.ResMap { //nolint:ireturn // kustomize API returns ResMap interface //nolint:ireturn
```
Comment (GowthamShanmugam) [inline internal/controller/components/modelsasservice/modelsasservice_support_test.go]: is this change required?

File: pkg/controller/reconciler/reconciler_finalizer_test.go [hunk 10]
```diff
@@ -56,21 +56,17 @@ func (f *MockManager) GetScheme() *runtime.Scheme { return f.scheme }
 func (f *MockManager) GetRESTMapper() meta.RESTMapper { return f.mapper }
 func (f *MockManager) GetConfig() *rest.Config        { return &rest.Config{} }
 
-//nolint:ireturn // Returns stdlib interface required by manager.Manager
-func (f *MockManager) GetFieldIndexer() client.FieldIndexer { return nil }
+func (f *MockManager) GetFieldIndexer() client.FieldIndexer { return nil } //nolint:ireturn // manager.Manager
```
Comment (GowthamShanmugam) [inline pkg/controller/reconciler/reconciler_finalizer_test.go]: is this changes are required?

File: internal/controller/services/gateway/gateway_tls_test.go (hunk 5 above)
Comment (GowthamShanmugam) [reply internal/controller/services/gateway/gateway_tls_test.go]: @sukumars321 please check this

File: internal/controller/services/gateway/gateway_tls.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2023.
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls.go]:
```suggestion
Copyright 2026.
```

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 12]
```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2023.
```
Comment (GowthamShanmugam) [inline internal/controller/services/gateway/gateway_tls_test.go]:
```suggestion
Copyright 2026.
```

File: internal/controller/services/gateway/gateway_tls.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+		return ""
+	}
+}
+func TLSCipherSuitesFromProfileSpec(spec *configv1.TLSProfileSpec) string {
+	if spec == nil {
+		// default to intermediate profile
+		spec = configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
+	}
+	ianaCiphers := ocpcrypto.OpenSSLToIANACipherSuites(spec.Ciphers)
+	return strings.Join(ianaCiphers, ",")
```
Comment (ugiordan) [inline internal/controller/services/gateway/gateway_tls.go:109]: Two issues here:

1. `OpenSSLToIANACipherSuites` silently drops DHE ciphers that Go's `crypto/tls` doesn't support. Intermediate profile loses 2 of 11 ciphers, Old loses 8 of 29. The enforced policy diverges from the cluster config with no warning. We should log when ciphers are dropped.

2. If all ciphers fail to map or the Custom profile has an empty Ciphers slice, this returns an empty string, producing `--tls-cipher-suite=` in the deployment template. Add a fallback to Intermediate ciphers when the result is empty, same pattern as the version floor in `tlsMinVersionFromProfileSpec`.

**Review by ugiordan — CHANGES_REQUESTED** [review]

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+		})
+	}
+}
+
+func TestKubeAuthProxyTLSFromProfile(t *testing.T) {
+	t.Parallel()
+
+	intermediateSpec := configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
+	expectedMin := TLSMinVersionToName(intermediateSpec.MinTLSVersion)
+	expectedCiphers := TLSCipherSuitesFromProfileSpec(intermediateSpec)
```
Comment (ugiordan) [inline internal/controller/services/gateway/gateway_tls_test.go]: This derives the expected value from the same production function being tested. If `TLSCipherSuitesFromProfileSpec` has a bug (say it swaps two ciphers, or drops one), the test computes the same wrong expected value and still passes.

Same pattern at line 142, and in [tests/e2e/gateway_tls_test.go:22](https://github.com/sukumars321/opendatahub-operator/blob/4c02b1bfe2f3e5288f749bbe4329d077d7a1a9a3/tests/e2e/gateway_tls_test.go#L22) where `kubeAuthProxyTLSDeploymentArgs` calls `gateway.KubeAuthProxyTLSFromProfile` to build expected deployment args.

The expected values should be hardcoded strings, e.g.:
```
    expectedCiphers := "TLS_AES_128_GCM_SHA256,TLS_AES_256_GCM_SHA384,..."
    expectedMin := "TLS1.2"
```

That way, the test is an independent oracle: production code computes the actual, the test asserts against a known-good constant.

**Review by ugiordan — CHANGES_REQUESTED** [review]

File: internal/controller/services/gateway/gateway_tls.go [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+func TLSMinVersionToName(minVersion configv1.TLSProtocolVersion) string {
+	switch minVersion {
+	case configv1.VersionTLS12:
+		return "TLS1.2"
+	case configv1.VersionTLS13:
+		return "TLS1.3"
+	default:
+		return ""
+	}
+}
```
Comment (ugiordan) [inline internal/controller/services/gateway/gateway_tls.go:78]:
```suggestion
}

```

Comment (ugiordan) [inline internal/controller/services/gateway/gateway_tls.go]: When an unsupported `MinTLSVersion` hits the default case (e.g., from the Old profile), the version gets floored to TLS 1.2. But the cipher list still comes from whatever profile was selected. The Old profile includes weak ciphers like `DES-CBC3-SHA` (maps to `TLS_RSA_WITH_3DES_EDE_CBC_SHA`). So you end up with "TLS 1.2 minimum, but accepting 3DES". If the version floor kicks in, the cipher list should probably be tightened to at least the Intermediate baseline too.

**Review by ugiordan — CHANGES_REQUESTED** [review]

File: internal/controller/services/gateway/gateway_tls.go (hunk 15 above)
Comment (ugiordan) [reply internal/controller/services/gateway/gateway_tls.go]: For example, 3DES is considered broken (Sweet32 attack, CVE-2016-2183). An attacker can't downgrade to TLS 1.0, but they CAN negotiate TLS 1.2 with a 3DES cipher suite, which is still weak.

The fix would be: if the version floor activates (the default case returns ""), also replace the cipher list with the Intermediate profile's ciphers. That way, the floor applies consistently to both versions and ciphers.

File: tests/e2e/gateway_tls_test.go [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+
+		args := deployment.Spec.Template.Spec.Containers[0].Args
+		g.Expect(args).To(And(
+			ContainElement(minVersionArg),
+			ContainElement(cipherSuitesArg),
+		))
+	}).WithTimeout(tc.TestTimeouts.defaultEventuallyTimeout).
+		WithPolling(tc.TestTimeouts.defaultEventuallyPollInterval).
+		Should(Succeed())
+}
```
[the remark this answers is not in the document]
Comment (ugiordan) [reply tests/e2e/gateway_tls_test.go:124]: this is a valid nit

**Comment by sukumars321** [comment]
@ugiordan, Addressed all the comments, PTAL.

File: internal/controller/services/gateway/gateway_tls.go (hunk 15 above)
Comment (sukumars321) [reply internal/controller/services/gateway/gateway_tls.go]: Addressed. Now we are flooring to intermediate profile that defaults to TLS1.2 and default ciphers.

File: internal/controller/services/gateway/gateway_tls_test.go (hunk 14 above)
Comment (sukumars321) [reply internal/controller/services/gateway/gateway_tls_test.go]: Addressed.

**Review by ugiordan — DISMISSED** [review]

File: internal/controller/services/gateway/gateway_tls_test.go [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+	// nil profile defaults to Intermediate; Old's TLS 1.0 floors to TLS 1.2
+	assert.Equal(t, "TLS1.2", minVersion)
+	assert.Equal(t, intermediateIANACiphers, cipherSuites)
+
+	minVersion, cipherSuites = KubeAuthProxyTLSFromProfile(context.Background(), &configv1.TLSSecurityProfile{Type: configv1.TLSProfileOldType})
+	// Old profile specifies TLS 1.0; both version AND ciphers floor to Intermediate
+	// so that weak Old ciphers (e.g. 3DES) are not paired with a TLS 1.2 minimum.
+	assert.Equal(t, "TLS1.2", minVersion)
+	assert.Equal(t, intermediateIANACiphers, cipherSuites)
+}
```
Comment (ugiordan) [inline internal/controller/services/gateway/gateway_tls_test.go:190]: nit: `TestKubeAuthProxyTLSFromProfile` covers nil and Old profiles but not Modern (TLS 1.3). A test case with `TLSProfileModernType` would exercise the `"TLS1.3"` output branch of `tlsMinVersionToName`, which currently has zero assertion coverage at the public API level.

File: pkg/controller/predicates/resources/resources_test.go [hunk 18]
```diff
@@ -1257,3 +1259,49 @@ func TestHTTPRouteReferencesGateway_DefaultNamespace(t *testing.T) {
...
+			ObjectOld: &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: cluster.ClusterAPIServerObj}},
+			ObjectNew: clusterAPIServer(configv1.TLSProfileOldType),
+		})).To(BeTrue(), "old object conversion failure should trigger reconcile")
+
+		g.Expect(predicate.Update(event.UpdateEvent{
+			ObjectOld: clusterAPIServer(configv1.TLSProfileIntermediateType),
+			ObjectNew: &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: cluster.ClusterAPIServerObj}},
+		})).To(BeTrue(), "new object conversion failure should trigger reconcile")
+	})
+}
```
Comment (ugiordan) [inline pkg/controller/predicates/resources/resources_test.go:1384]: nit: `TestAPIServerTLSSecurityProfileChanged` only tests UpdateFunc. CreateFunc, DeleteFunc, and GenericFunc (which always return false) have no dedicated assertions. Could add a quick table-driven check for completeness.

**Review by ugiordan — COMMENTED** [review]
Minor suggestions, not blocking.

**Review by ugiordan — APPROVED** [review]
Both suggestions addressed: Modern/TLS1.3 test added, predicate tests expanded to cover Create/Delete/Generic and edge cases. LGTM.

**Review by GowthamShanmugam — APPROVED** [review]

**Comment by GowthamShanmugam** [comment]
e2e is keep restarting, but it has a pass history(https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-opendatahub-io-opendatahub-operator-main-opendatahub-operator-rhoai-e2e). i am overriding.

**Comment by GowthamShanmugam** [comment]
/override "Red Hat Konflux / odh-operator-ci-e2e-its-manual / odh-operator-ci"

**Comment by GowthamShanmugam** [comment]
/override "Red Hat Konflux / odh-operator-ci-on-pull-request"

**Comment by GowthamShanmugam** [comment]
/override "Red Hat Konflux / odh-operator-ci-e2e-its-manual / odh-operator-ci"

**Comment by GowthamShanmugam** [comment]
/override "Red Hat Konflux / odh-operator-ci-on-pull-request"
